File: Platform/Workspace/Typescript/src/allors/framework/database/pull/Predicate.ts
import { ObjectType, RoleType, PropertyType } from '../../meta';
import { Extent } from './Extent';
import { PredicateKind } from '../data/PredicateKind';

export class Predicate {

  public kind: PredicateKind;

  public propertyType: PropertyType;

  public roleType: RoleType;

  public objectType: ObjectType;

  public parameter: string;

  public operand: Predicate; 

  public operands: Predicate[];

  public object: string;

  public objects: string[];

  public value: string;

  public values: string[];

  public Extent: Extent;

  constructor(fields?: Partial<Predicate>) {
    Object.assign(this, fields);
  }

  public toJSON(): any {

    return {
      kind: this.kind,
      objectType: this.objectType.id,
    };
  }
}

File: Platform/Framework/Allors.Protocol/Data/Predicate.cs
// <copyright file="Predicate.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Protocol.Data
{
    using System;

    public class Predicate
    {
        public PredicateKind Kind { get; set; }

        public Guid? PropertyType { get; set; }

        public Guid? RoleType { get; set; }

        public Guid? ObjectType { get; set; }

        public string Parameter { get; set; }

        public Predicate Operand { get; set; }

        public Predicate[] Operands { get; set; }

        public string Object { get; set; }

        public string[] Objects { get; set; }

        public string Value { get; set; }

        public string[] Values { get; set; }

        public Extent Extent { get; set; }
    }
}
